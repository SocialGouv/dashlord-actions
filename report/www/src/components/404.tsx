import * as React from "react";

import Table from "@codegouvfr/react-dsfr/Table";

import { Panel } from "./Panel";

type Report404Props = { data: Wget404Report };

export const Report404: React.FC<Report404Props> = ({ data }) => {
  const tableData = data && [
    ["destination"],
    ...data.map((url2) => [url2.link]),
  ];
  return (
    (data && data.length && (
      <Panel title="Erreurs 404" info="Pages introuvables">
        <Table data={tableData} />
      </Panel>
    )) ||
    null
  );
};
