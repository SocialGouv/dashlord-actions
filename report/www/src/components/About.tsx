import * as React from "react";

import CallOut from "@codegouvfr/react-dsfr/CallOut";

export const About: React.FC = () => (
  <CallOut className="fr-mb-3w" title="À propos">
    DashLord est <strong>100% open source</strong>, vous pouvez y contribuer en
    créant de la documentation, en proposant du code, ou encore en partageant
    vos questions{" "}
    <a
      href="https://github.com/socialgouv/dashlord"
      target="_blank"
      rel="noopener noreferrer"
    >
      sur le repository github
    </a>
    .
    <br /> <br />
    L&apos;objectif de DashLord est d&apos;identifier des points
    d&apos;attention mais aussi de proposer des solutions actionnables.
    <br /> <br />
    DashLord est né à la{" "}
    <a
      href="https://fabrique.social.gouv.fr"
      target="_blank"
      rel="noopener noreferrer"
    >
      {" "}
      Fabrique des ministères sociaux
    </a>{" "}
    pour répondre aux besoins d&apos;évaluation et de mise en oeuvre des bonnes
    pratiques de développement web.
    <br /> <br />
    Vous pouvez facilement créer votre propre DashLord pour vos URLs en suivant{" "}
    <a
      href="https://github.com/SocialGouv/dashlord/blob/main/README.md"
      target="_blank"
      rel="noopener noreferrer"
    >
      les instructions sur le GitHub
    </a>
    .
  </CallOut>
);
