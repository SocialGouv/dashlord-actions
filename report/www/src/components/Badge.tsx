import classNames from "classnames";
import React, { CSSProperties } from "react";
import { useRouter } from "next/router";

import styles from "./badge.module.scss";

type BadgeProps = {
  variant: string;
  className?: string;
  to?: string;
  style?: CSSProperties;
  external?: boolean;
  children: any;
};

const Badge: React.FC<BadgeProps> = ({
  children,
  external = false,
  variant,
  className,
  to,
  style,
}) => {
  const router = useRouter();
  return (
    <button
      type="button"
      className={classNames(className, styles[variant])}
      onClick={() => {
        if (to) {
          if (external) {
            window.open(to);
            return;
          }
          router.push(to);
        }
      }}
      style={style}
    >
      {children}
    </button>
  );
};

export default Badge;
