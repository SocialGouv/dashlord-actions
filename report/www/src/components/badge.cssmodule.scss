.success,
.info,
.warning,
.danger {
    display: inline-block;
    cursor: pointer;
    padding: 4px 8px;
    color: white;
    min-width: 60px;
    border-radius: 4px;
    text-align: center;
    font-weight: bold;
}

.success {
    background-color: var(--success);
}

.info {
    background-color: var(--info);
}

.warning {
    background-color: #FF6F4C;
}

.danger {
    background-color: var(--error);
}