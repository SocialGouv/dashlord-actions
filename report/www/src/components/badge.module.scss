.success,
.info,
.warning,
.danger {
    display: inline-block;
    cursor: pointer;
    padding: 4px 8px;
    color: white;
    min-width: 60px;
    border-radius: 4px;
    text-align: center;
    font-weight: bold;
}



.success, .success:hover {
    background-color: var(--success) !important;
}

.info, .info:hover {
    background-color: var(--info) !important;
}

.warning, .warning:hover {
    background-color: var(--warning) !important;
}

.danger, .danger:hover {
    background-color: var(--error) !important;
}