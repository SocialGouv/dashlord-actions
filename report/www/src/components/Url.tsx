import * as React from "react";
import { Info, Zap, ThumbsUp, Lock } from "react-feather";

import { isToolEnabled, slugifyUrl, btoa } from "../utils";
import { HTTP } from "./HTTP";
import { LightHouse } from "./LightHouse";
import { Nuclei } from "./Nuclei";
import { Owasp } from "./Owasp";
import { TestSSL } from "./TestSSL";
import { Trackers } from "./Trackers";
import { Wappalyzer } from "./Wappalyzer";
import { UpdownIo } from "./UpdownIo";
import { Dependabot } from "./Dependabot";
import { Codescan } from "./Codescan";
import { Nmap } from "./Nmap";
import { Report404 } from "./404";
import { Trivy } from "./Trivy";
import { DeclarationA11y } from "./DeclarationA11y";
import { DeclarationRgpd } from "./DeclarationRgpd";
import { Tab, TabContent } from "./UrlTabs";
import { UrlHeader } from "./UrlHeader";
import { Panel } from "./Panel";
import { Page } from "./Page";
import { Betagouv } from "./BetagouvInfo";
import { GithubRepository } from "./GithubRepository";
import { EcoIndex } from "./EcoIndex";
import { SonarCloud } from "./SonarCloud";

const BASE_PATH = process.env.NEXT_PUBLIC_BASE_PATH || "";

type UrlDetailProps = { url: string; report: UrlReport; selectedTab?: string };

const Anchor = ({ id }: { id: string }) => <div id={id} />;

// define tabs structure
const tabs = [
  {
    label: "Bonnes pratiques",
    id: "best-practices",
    icon: <ThumbsUp size={16} style={{ marginRight: 5, marginBottom: -2 }} />,
    items: [
      {
        id: "lighthouse",
        reportKey: "lhr",
        render: (report, url) => (
          <LightHouse
            data={report.lhr}
            url={`${BASE_PATH}/report/${btoa(url)}`}
          />
        ),
      },
      {
        id: "ecoindex",
        reportKey: "ecoindex",
        render: (report, url) => <EcoIndex data={report.ecoindex} />,
      },
      {
        id: "thirdparties",
        render: (report) => <Trackers data={report.thirdparties} />,
      },
      {
        id: "stats",
        render: (report, url) =>
          report.stats ? (
            <Panel
              title="Page de statistiques"
              info="Cette page permet de publier vos mesures d'impact"
            >
              <Page data={report.stats} url={url} uri="stats" />
            </Panel>
          ) : null,
      },
      {
        id: "budget_page",
        render: (report, url) =>
          report.budget_page ? (
            <Panel
              title="Page de budget"
              info="Cette page permet de publier votre budget"
            >
              <Page data={report.budget_page} url={url} uri="budget" />
            </Panel>
          ) : null,
      },
      {
        id: "declaration-a11y",
        render: (report, url) => (
          <DeclarationA11y data={report["declaration-a11y"]} />
        ),
      },
      {
        id: "declaration-rgpd",
        render: (report, url) => (
          <DeclarationRgpd data={report["declaration-rgpd"]} />
        ),
      },
      {
        id: "404",
        render: (report, url) => <Report404 data={report["404"]} />,
      },
      {
        id: "github_repository",
        render: (report) => (
          <GithubRepository data={report["github_repository"]} />
        ),
      },
    ],
  },
  {
    label: "Disponibilité",
    id: "disponibilite",
    icon: <Zap size={16} style={{ marginRight: 5, marginBottom: -2 }} />,
    items: [
      {
        id: "updownio",
        render: (report, url) => <UpdownIo data={report.updownio} url={url} />,
      },
    ],
  },
  {
    label: "Sécurité",
    id: "securite",
    icon: <Lock size={16} style={{ marginRight: 5, marginBottom: -2 }} />,
    items: [
      {
        id: "nmap",
        render: (report, url) => (
          <Nmap
            data={report.nmap}
            url={`${BASE_PATH}/report/${btoa(url)}/nmapvuln.html`}
          />
        ),
      },
      {
        id: "http",
        render: (report, url) => <HTTP data={report.http} />,
      },
      {
        id: "testssl",
        render: (report, url) => (
          <TestSSL
            data={report.testssl}
            url={`${BASE_PATH}/report/${btoa(url)}/testssl.html`}
          />
        ),
      },
      {
        id: "dependabot",
        render: (report, url) =>
          report.dependabot.repositories &&
          report.dependabot.repositories
            .filter(Boolean)
            .map((repository) => (
              <Dependabot key={repository.url} data={repository} url={url} />
            )),
      },
      {
        id: "codescan",
        render: (report, url) =>
          report.codescan.repositories &&
          report.codescan.repositories
            .filter(Boolean)
            .map((repository) => (
              <Codescan key={repository.url} data={repository} url={url} />
            )),
      },
      {
        id: "zap",
        render: (report, url) => (
          <Owasp
            data={report.zap}
            url={`${BASE_PATH}/report/${btoa(url)}/zap.html`}
          />
        ),
      },
      {
        id: "nuclei",
        render: (report, url) => <Nuclei data={report.nuclei} />,
      },
      {
        id: "trivy",
        render: (report, url) =>
          report["trivy"].length && (
            <Trivy
              data={report["trivy"]}
              url={`${BASE_PATH}/report/${btoa(url)}/trivy.json`}
            />
          ),
      },
      {
        id: "sonarcloud",
        render: (report, url) =>
          report["sonarcloud"].length && (
            <SonarCloud data={report["sonarcloud"]} />
          ),
      },
    ],
  },
  {
    label: "Informations",
    id: "informations",
    icon: <Info size={16} style={{ marginRight: 5, marginBottom: -2 }} />,
    items: [
      {
        id: "wappalyzer",
        render: (report) => <Wappalyzer data={report.wappalyzer} />,
      },
      {
        id: "betagouv",
        render: (report) => <Betagouv data={report.betagouv} />,
      },
    ],
  },
];

export const Url: React.FC<UrlDetailProps> = ({ url, report, selectedTab }) => {
  React.useEffect(() => {
    const hash = document.location.hash.split("#");
    if (hash.length === 3) {
      // double hash + HashRouter workaround
      const target = document.getElementById(hash[2]);
      if (target) {
        target.scrollIntoView();
      }
    }
  }, [report]);

  if (!report) {
    return (
      <div>
        No data available for
        {url}
      </div>
    );
  }

  const selectedTabIndex: number = Math.max(
    0,
    tabs.findIndex((tab) => selectedTab === tab.id)
  );

  const selectedTabDefinition = tabs[selectedTabIndex];

  return (
    <>
      <UrlHeader report={report} url={url} />

      {/* custom DSFR Tabs renderer because of some react-dsfr issue */}
      <div className="fr-tabs">
        <ul className="fr-tabs__list" role="tablist">
          {tabs.map((tab, tabIndex) => (
            <Tab
              key={tab.id}
              selected={selectedTabDefinition.id === tab.id}
              index={tabIndex}
              href={`/url/${slugifyUrl(url)}/${tab.id}`}
              {...tab}
            />
          ))}
        </ul>
        {tabs.map((tab, tabIndex) => {
          // filter out invalid items
          const items = tab.items
            .filter(
              (item) =>
                !!report[item.reportKey || item.id] &&
                isToolEnabled(item.id as DashlordTool)
            )
            .map((item) => (
              <div key={item.id}>
                <Anchor id={item.id} />
                {item.render(report, url)}
              </div>
            ));

          return (
            <TabContent
              {...tab}
              key={tab.id}
              tabIndex={tabIndex}
              selected={selectedTabDefinition.id === tab.id}
              items={items}
            />
          );
        })}
      </div>
    </>
  );
};
