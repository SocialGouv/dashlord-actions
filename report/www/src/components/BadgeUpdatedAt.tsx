import { formatDistanceToNow, differenceInDays, format } from "date-fns";
import { fr } from "@codegouvfr/react-dsfr";
import frLocale from "date-fns/locale/fr";

import { AlertProps } from "@codegouvfr/react-dsfr/Alert";
import Badge from "@codegouvfr/react-dsfr/Badge";
import { CSSProperties } from "react";

// badge that displya "since xxx" and appropriate colors
export const BadgeUpdatedAt = ({
  date,
  style,
}: {
  date: string;
  style: CSSProperties;
}) => {
  if (!date) return;
  const startDate = new Date(date);
  const lagInDays = differenceInDays(new Date(), startDate);
  let severity: AlertProps.Severity = "info";
  if (lagInDays > 60) {
    severity = "error";
  } else if (lagInDays > 30) {
    severity = "warning";
  }
  return (
    <Badge
      severity={severity}
      noIcon
      as="span"
      style={{ verticalAlign: "top", ...style }}
    >
      <i
        className={fr.cx("fr-icon-time-fill", "fr-icon--sm", "fr-mr-1v")}
        title={`Mis Ã  jour le ${format(new Date(startDate), "dd/M/yyyy")}`}
      />
      {formatDistanceToNow(new Date(startDate), {
        locale: frLocale,
      })}
    </Badge>
  );
};
