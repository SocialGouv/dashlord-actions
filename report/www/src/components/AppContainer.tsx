import * as React from "react";

import { Container, Alert } from "@dataesr/react-dsfr";

import { Switch, Route, useLocation, useParams } from "react-router-dom";

import { Dashboard } from "./Dashboard";
import { Trends } from "./Trends";
import { Intro } from "./Intro";
import { About } from "./About";
import { WappalyzerDashboard } from "./WappalyzerDashboard";
import Url from "./Url";

const report: DashLordReport = require("./report.json");
const trends: Trends = require("./trends.json");

type CategoryRouteProps = { report: DashLordReport };

// for some reason react-router `:url*` didnt work, use `*` only
interface CategoryParamTypes {
  category: string;
}

const CategoryRoute: React.FC<CategoryRouteProps> = (props) => {
  const params = useParams<CategoryParamTypes>();
  const category = window.decodeURIComponent(params.category);
  const urls = props.report.filter((u) => u.category === category);
  return (
    <>
      <br />
      {urls.length ? (
        <>
          <h3>
            {category} :{urls.length} urls
          </h3>
          <Dashboard report={urls} />
        </>
      ) : (
        <h3>Aucun URL associ√©e</h3>
      )}
    </>
  );
};

interface TagParamTypes {
  tag: string;
}

type TagRouteProps = { report: DashLordReport };

const TagRoute: React.FC<TagRouteProps> = (props) => {
  const params = useParams<TagParamTypes>();
  const tag = window.decodeURIComponent(params.tag);
  const urls = props.report.filter((u) => u.tags && u.tags.includes(tag));
  return (
    <>
      <br />
      <h3>
        {tag} :{urls.length} urls
      </h3>
      <Dashboard report={urls} />
    </>
  );
};

interface UrlParamTypes {
  "0": string;
}

type UrlRouteProps = { report: DashLordReport };

const UrlRoute: React.FC<UrlRouteProps> = (props) => {
  const params = useParams<UrlParamTypes>();
  const url = window.decodeURIComponent(params["0"]);
  const urlData = props.report.find((u) => u.url === url);
  if (!urlData) {
    return (
      <Alert
        type="error"
        title={`Impossible de trouver le rapport pour ${url}`}
      />
    );
  }
  return <Url url={url} report={urlData} />;
};

export const AppContainer = () => {
  const location = useLocation();
  return (
    <Container fluid={location.pathname === "/"}>
      <div role="main" className="fr-my-4w">
        <Switch>
          <Route path="/url/*">
            <UrlRoute report={report} />
          </Route>
          <Route path="/dashboard">
            <Dashboard report={report} />
          </Route>
          <Route path="/trends">
            <Trends trends={trends} />
          </Route>
          <Route path="/category/:category">
            <CategoryRoute report={report} />
          </Route>
          <Route path="/tag/:tag">
            <TagRoute report={report} />
          </Route>
          <Route path="/wappalyzer">
            <WappalyzerDashboard report={report} />
          </Route>
          <Route path="/intro">
            <Intro />
          </Route>
          <Route path="/about">
            <About />
          </Route>
          <Route path="/">
            <Dashboard report={report} />
          </Route>
        </Switch>
      </div>
    </Container>
  );
};
