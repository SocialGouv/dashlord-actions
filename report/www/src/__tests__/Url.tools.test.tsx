import * as renderer from "react-test-renderer";

import { Url } from "@/components/Url";
import { vi } from "vitest";

const date = new Date(2021, 3, 6);

vi.useFakeTimers();
vi.setSystemTime(date);

const TEST_URL = "https://www.fabrique.social.gouv.fr";

const report = require("../report.json").find((r) => r.url === TEST_URL);

describe("Tools config", () => {
  beforeAll(() => {
    vi.mock("@/config.json", async () => {
      const config = {
        ...(await vi.importActual("@/config.json")),
        tools: { http: false, zap: true, lighthouse: false },
      };
      return {
        default: config,
      };
    });
  });

  it("Should render Url with limited tools", () => {
    const props = { report, url: TEST_URL };
    const tree = renderer.create(<Url {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
