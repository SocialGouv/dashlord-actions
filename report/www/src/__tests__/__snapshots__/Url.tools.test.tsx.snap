// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tools config > Should render Url with limited tools 1`] = `
[
  <div
    className="fr-callout fr-mb-3w"
    id="fr-callout-:r0:"
  >
    <h3
      className="fr-callout__title"
    >
      <a
        href="https://www.fabrique.social.gouv.fr"
        rel="noreferrer noopener"
        target="_blank"
      >
        https://www.fabrique.social.gouv.fr
      </a>
    </h3>
    <p
      className="fr-callout__text"
    >
       
      <div
        className="fr-mt-1w"
      >
        <span
          className="fr-badge fr-badge--warning fr-badge--no-icon"
          id="fr-badge-:r1:"
          style={
            {
              "verticalAlign": "top",
            }
          }
        >
          <i
            className="fr-icon-time-fill fr-icon--sm fr-mr-1v"
            title="Mis à jour il y a environ 1 an"
          />
          environ 1 an
        </span>
        <a
          className="link-discreet"
          href="/category/fabrique"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
          title="Voir toutes les URLs de la catégorie fabrique"
        >
          <p
            className="fr-badge fr-badge--success"
            id="fr-badge-:r2:"
          >
            fabrique
          </p>
        </a>
        <a
          className="link-discreet"
          href="/tag/social"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
          title="Voir toutes les URLs du tag social"
        >
          <p
            className="fr-badge fr-badge--info"
            id="fr-badge-:r3:"
          >
            social
          </p>
        </a>
        <a
          className="link-discreet"
          href="/tag/sante"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
          title="Voir toutes les URLs du tag sante"
        >
          <p
            className="fr-badge fr-badge--info"
            id="fr-badge-:r4:"
          >
            sante
          </p>
        </a>
        <a
          className="link-discreet"
          href="/tag/travail"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
          title="Voir toutes les URLs du tag travail"
        >
          <p
            className="fr-badge fr-badge--info"
            id="fr-badge-:r5:"
          >
            travail
          </p>
        </a>
      </div>
       
    </p>
  </div>,
  <div
    className="fr-tabs"
    id="fr-tabs-:r6:"
  >
    <ul
      className="fr-tabs__list"
      onKeyDownCapture={[Function]}
      role="tablist"
    >
      <li
        role="presentation"
      >
        <button
          aria-controls="tabpanel-:r7:-0-panel"
          aria-selected={true}
          className="fr-tabs__tab fr-icon-thumb-up-line fr-tabs__tab--icon-left"
          id="tabpanel-:r7:-0"
          onClick={[Function]}
          role="tab"
          tabIndex={0}
        >
          Bonnes pratiques
        </button>
      </li>
      <li
        role="presentation"
      >
        <button
          aria-controls="tabpanel-:r7:-1-panel"
          aria-selected={false}
          className="fr-tabs__tab fr-icon-time-line fr-tabs__tab--icon-left"
          id="tabpanel-:r7:-1"
          onClick={[Function]}
          role="tab"
          tabIndex={-1}
        >
          Disponibilité
        </button>
      </li>
      <li
        role="presentation"
      >
        <button
          aria-controls="tabpanel-:r7:-2-panel"
          aria-selected={false}
          className="fr-tabs__tab fr-icon-lock-line fr-tabs__tab--icon-left"
          id="tabpanel-:r7:-2"
          onClick={[Function]}
          role="tab"
          tabIndex={-1}
        >
          Sécurité
        </button>
      </li>
      <li
        role="presentation"
      >
        <button
          aria-controls="tabpanel-:r7:-3-panel"
          aria-selected={false}
          className="fr-tabs__tab fr-icon-info-line fr-tabs__tab--icon-left"
          id="tabpanel-:r7:-3"
          onClick={[Function]}
          role="tab"
          tabIndex={-1}
        >
          Informations
        </button>
      </li>
    </ul>
    <div
      aria-labelledby="tabpanel-:r7:-0"
      className="fr-tabs__panel fr-tabs__panel--selected"
      id="tabpanel-:r7:-0-panel"
      role="tabpanel"
      tabIndex={0}
    >
      <div
        className="fr-alert fr-alert--error"
        id="fr-alert-:r8:"
      >
        <h3
          className="fr-alert__title"
        >
          warn-no-info
        </h3>
        <div>
          Aucune information trouvée dans cette catégorie
        </div>
      </div>
    </div>
  </div>,
]
`;
